internal/provider/resources/resource_api_token.go

```diff
@@ -0,0 +1,566 @@
+		return
+	}
+
+	// delete request
+	apiToken, err := r.IamClient.DeleteApiTokenWithResponse(
+		ctx,
+		r.OrganizationId,
+		data.Id.ValueString(),
+	)
+	if err != nil {
+		tflog.Error(ctx, "failed to delete API token", map[string]interface{}{"error": err})
+		resp.Diagnostics.AddError(
+			"Client Error",
+			fmt.Sprintf("Unable to delete API token, got error: %s", err),
+		)
+		return
+	}
+	statusCode, diagnostic := clients.NormalizeAPIError(ctx, apiToken.HTTPResponse, apiToken.Body)
+	// It is recommended to ignore 404 Resource Not Found errors when deleting a resource
+	if statusCode != http.StatusNotFound && diagnostic != nil {
+		resp.Diagnostics.Append(diagnostic)
+		return
+	}
+
+	tflog.Trace(ctx, fmt.Sprintf("deleted an API token resource: %v", data.Id.ValueString()))
+}
+
+func (r *ApiTokenResource) ImportState(
+	ctx context.Context,
+	req resource.ImportStateRequest,
+	resp *resource.ImportStateResponse,
+) {
+	resource.ImportStatePassthroughID(ctx, path.Root("id"), req, resp)
+}
+
+func (r *ApiTokenResource) ValidateConfig(
+	ctx context.Context,
+	req resource.ValidateConfigRequest,
+	resp *resource.ValidateConfigResponse,
+) {
+	var data models.ApiTokenResource
+
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Convert Terraform set of roles to API token roles
+	roles, diags := RequestApiTokenRoles(ctx, data.Roles)
+	if diags.HasError() {
+		resp.Diagnostics.Append(diags...)
+		return
+	}
+
+	tokenRole, _, diags := RequestApiTokenRole(roles, data.Type.ValueString())
+	if diags != nil {
+		resp.Diagnostics.Append(diags...)
+		return
+	}
+
+	entityType := data.Type.ValueString()
+
+	// Check if the role is valid for the entity type
+	if !utils.CheckRole(tokenRole, entityType) {
+		resp.Diagnostics.AddError(
+			"Client Error",
+			fmt.Sprintf("Role %s is not valid for entity type %s", tokenRole, entityType),
+		)
+		return
+	}
+
+	// Validate the roles based on the entity type
+	switch entityType {
+	case string(iam.ApiTokenRoleEntityTypeORGANIZATION):
+		resp.Diagnostics.Append(validateOrganizationApiToken(roles)...)
+	case string(iam.ApiTokenRoleEntityTypeWORKSPACE):
+		resp.Diagnostics.Append(validateWorkspaceApiToken(roles)...)
+	case string(iam.ApiTokenRoleEntityTypeDEPLOYMENT):
+		resp.Diagnostics.Append(validateDeploymentApiToken(roles)...)
+	}
+}
+
+func validateOrganizationApiToken(roles []iam.ApiTokenRole) diag.Diagnostics {
+	for _, role := range roles {
+		if utils.CheckRole(role.Role, string(role.EntityType)) {
+			return nil
+		}
+	}
+	return diag.Diagnostics{
+		diag.NewErrorDiagnostic(
+			"Client Error",
+			"Unable to find the role for the entity type",
+		),
+	}
+}
+
+func validateWorkspaceApiToken(roles []iam.ApiTokenRole) diag.Diagnostics {
+	for _, role := range roles {
+		if role.EntityType == iam.ApiTokenRoleEntityTypeORGANIZATION {
+			return diag.Diagnostics{
+				diag.NewErrorDiagnostic(
+					"Client Error",
```

Comment:
All of these should be `Bad Request Error` with proper descriptions
`Client Error` is when the API call to core errors